### PR TITLE
Feature/unstaking subscription

### PR DESCRIPTION
**Pull request by OrLavy**
## Description

Describe the big picture of your changes here to communicate to the maintainers why we should accept this pull request. 
If it fixes a `bug` or resolves a `feature request`, be sure to link to that issue.

## Types of changes

What types of changes does your code introduce?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code._

- [ ] I have read the [CONTRIBUTING](https://github.com/orbs-network/orbs-contributor-guide/blob/master/CONTRIBUTING.md) doc
- [ ] I have opened an issue and discussed the solution suggested in this PR
- [ ] Lint and unit tests pass locally with my changes
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: voting/orbs-pos-data/src/__tests__/staking-service-mock.test.ts [hunk 1]
```diff
@@ -8,11 +8,13 @@
 import { StakingServiceMock } from '../testkit';
 import { IStakingService } from '../interfaces/IStakingService';
 import { testTxCreatingForServiceMock } from './testUtils/txCreatingMethodTests';
+import 'jest-expect-message';
 
 describe(`Staking service mock`, () => {
   testTxCreatingMethods();
-  testDataReadingMethods();
-  testSubscription();
+  testDataReadingMethodsWitchMockHelpers();
```
Comment (gilamran) [inline voting/orbs-pos-data/src/__tests__/staking-service-mock.test.ts]: with

File: voting/orbs-pos-data/src/__tests__/staking-service-mock.test.ts [hunk 2]
```diff
@@ -24,58 +26,137 @@ function testTxCreatingMethods() {
   });
 }
 
-function testDataReadingMethods() {
-  describe(`Data reading methods`, () => {
-    let stakingService: StakingServiceMock;
+function testDataReadingMethodsWitchRealAPI() {
```
Comment (gilamran) [inline voting/orbs-pos-data/src/__tests__/staking-service-mock.test.ts]: rename test that it imitates real functionality

File: voting/orbs-pos-data/src/__tests__/staking-service-mock.test.ts [hunk 3]
```diff
@@ -24,58 +26,137 @@ function testTxCreatingMethods() {
...
+
+      // Restake
+      await stakingService.restake();
+      const resultAfterRestake = await stakingService.readTotalStakedTokens();
+      expect(resultAfterRestake, 'Should add all restaked tokens to the total count').toEqual('666666');
+    });
+
+    it(`Should return the correct cooldown status after stake, unstake and restake`, async () => {
+      const cooldownPeriodTimeInSeconds = 100;
+      (stakingService as StakingServiceMock).setCooldownTime(cooldownPeriodTimeInSeconds);
```
Comment (gilamran) [inline voting/orbs-pos-data/src/__tests__/staking-service-mock.test.ts]: Start the mock with now+day, and the user of the testkit can override it with setNextCooldownTime in your tests

File: voting/orbs-pos-data/src/services/StakingService.ts [hunk 4]
```diff
@@ -6,14 +6,32 @@
...
+  StakingServiceEventCallback,
+} from '../interfaces/IStakingService';
 import { getUnsubscribePromise } from '../utils/erc20EventsUtils';
+import { ITypedEventData, TUnsubscribeFunction } from './contractsTypes/contractTypes';
+
+/**
+ * It just so happens that all of the staking related events have the same signature.
+ */
+interface IStakingContractEventValues {
+  1: string;
```
Comment (gilamran) [inline voting/orbs-pos-data/src/services/StakingService.ts]: remove array access

File: voting/orbs-pos-data/src/testkit/StakingServiceMock.ts [hunk 5]
```diff
@@ -1,55 +1,101 @@
...
 
   private stakingContractAddress: string = 'DUMMY_CONTRACT_ADDRESS';
-  private addressToBalanceMap: Map<string, number> = new Map();
-  private addressToStakeStatus: Map<string, IStakingStatus> = new Map();
-  private stakeAmountChangeEventsMap: Map<string, StakeAmountChangeCallback> = new Map();
+  private addressToTotalStakedAmountMap: Map<string, number> = new Map();
+  private addressToCooldownStatus: Map<string, IStakingStatus> = new Map();
   private totalStakedTokens: string = '0';
 
+  private stakedEventsMap: Map<string, Map<number, StakingServiceEventCallback>> = new Map();
```
Comment (gilamran) [inline voting/orbs-pos-data/src/testkit/StakingServiceMock.ts]: Class this

File: voting/orbs-pos-data/src/__tests__/staking-service-mock.test.ts (hunk 1 above)
Comment (OrLavy) [reply voting/orbs-pos-data/src/__tests__/staking-service-mock.test.ts]: Done

File: voting/orbs-pos-data/src/__tests__/staking-service-mock.test.ts (hunk 2 above)
Comment (OrLavy) [reply voting/orbs-pos-data/src/__tests__/staking-service-mock.test.ts]: Done

File: voting/orbs-pos-data/src/services/StakingService.ts (hunk 4 above)
Comment (OrLavy) [reply voting/orbs-pos-data/src/services/StakingService.ts]: Done

File: voting/orbs-pos-data/src/__tests__/staking-service-mock.test.ts (hunk 3 above)
Comment (OrLavy) [reply voting/orbs-pos-data/src/__tests__/staking-service-mock.test.ts]: Done

**Pull request merged**